### PR TITLE
Enhance Scenario Filters

### LINKED ISSUE #1511: The ability to filter and sort scenarios in scenario selector according to data node contents
### Description

Zach created a custom scenario selector that displays a summary table of your scenarios with the content of some of their data nodes.

This selector is helpful because it is easy to find the scenario I want from the list. I can also easily filter where "Promotion?" is True and sort by "Air Pollution" in descending order.

image

### Solution Proposed

The scenario selector visual element can only sort and filter according to tags. It would be helpful if the scenario selector had a property `filter_dn` associated with a list of data node names. These would be data node names on which the end user could filter and sort his scenarios. For example, it would be `<|{scenario}|scenario_selector|filter_dn=Promotion?;Air Pollution|>`.

### Impact of Solution

The idea here is to allow the user to filter and sort his scenarios without having to create callbacks to define proper tags. Also, since the dev defines which data nodes to filter, he can ensure that the user does not try to sort complex data structures.

### Additional Context

_No response_

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
resolves #1511

```
import datetime as dt

import taipy as tp
from taipy import Config, Frequency, Scenario
from taipy.gui_core import ScenarioCustomFilter, ScenarioDatanodeFilter, ScenarioFilter


# Function to run a Dataiku scenario
def run_something(input_1, input_2):
    datetime = dt.datetime.now()
    date = dt.date(2018, 1, 1)
    int_var = 10
    string_var = "String"
    return datetime, date, int_var, string_var


data = {"toto": list(range(10_000)), "titi": [2 * i for i in range(10_000)], "tata": [4 * i for i in range(10_000)]}


input_1_cfg = Config.configure_data_node(
    id="input_1_data_node",
    default_data=data,
)


input_2_cfg = Config.configure_data_node(
    id="input_2_data_node",
    default_data=data,
)

datetime_cfg = Config.configure_data_node(id="datetime_data_node")
date_cfg = Config.configure_data_node(id="date_data_node")
int_cfg = Config.configure_data_node(id="int_data_node")
string_cfg = Config.configure_data_node(id="string_data_node")


# Scenario and task configuration in Taipy
scenario_task_cfg = Config.configure_task(
    id="scenario_task",
    function=run_something,
    input=[input_1_cfg, input_2_cfg],
    output=[datetime_cfg, date_cfg, int_cfg, string_cfg],
)

scenario_cfg = Config.configure_scenario(id="scenario", task_configs=[scenario_task_cfg], frequency=Frequency.DAILY)

def customfilter(scenario: Scenario):
    return scenario.int_data_node.read() if scenario.int_data_node.is_valid else None

filter_for_scenario = [
    "*",  # get default list
    ScenarioFilter("Number of Warehouses", int, "Number_of_Warehouses"),  # on int propety
    ScenarioFilter("Fix warehouses", bool, "Fix_warehouses"),  # on bool property
    ScenarioFilter("Fixed warehouses", str, "Fixed_warehouses"),  # on str property
    ScenarioFilter("Warehouse Date", dt.datetime, "date_wh"),  # on date property
    DataNodeScenarioFilter(
        "int Datanode Last Edit Date", dt.datetime, "int_data_node", "last_edit_date"
    ),  # on datanode date property
    ScenarioFilter("int Datanode", int, "int_data_node"),  # on datanode value
    CustomScenarioFilter("custom filter", int, customfilter),  # on int custom filter value
    # CustomScenarioFilter("custom filter not supported", int, lambda s: s.toto),  # not supported
]
# filter_for_scenario = CustomScenarioFilter("custom filter", int, customfilter) # single filter

# GUI Markdown content
scenario_md = """
<|{scenario}|scenario_selector|filter={filter_for_scenario}|>

<|{scenario}|scenario|>
"""


def add_properties(scenario):
    scenario.properties["Fix_warehouses"] = True
    scenario.properties["Number_of_Warehouses"] = 5
    scenario.properties["Fixed_warehouses"] = ["Louisville"]
    scenario.properties["Fixed warehouses"] = ["Louisville"]
    scenario.properties["date_wh"] = dt.datetime.now()
    return scenario


# Main execution block with GUI setup
if __name__ == "__main__":
    tp.Core().run()
    scenario = tp.create_scenario(scenario_cfg)
    scenario = add_properties(scenario)

    tp.Gui(scenario_md).run(title="1511 ENhance Scenario Filters")

```

### CODE REVIEW

File: taipy/gui_core/viselements.json [hunk 1]
```diff
@@ -100,8 +100,8 @@
                     },
                     {
                         "name": "filter",
-                        "type": "bool|str|list[str]",
-                        "default_value": "\"Config id;Label;Creation date;Cycle label;Cycle start;Cycle end;Primary;Tags\"",
+                        "type": "bool|str|list[str|ScenarioFilter]",
```
Comment (FabienLelaquais) [inline taipy/gui_core/viselements.json]: Would a single filter make sense?
-> `bool|str|ScenarioFilter|list[str|ScenarioFilter]`

**Review by FabienLelaquais — DISMISSED** [review]
Nothing hurts my eyes.
Amazing stuff there!

File: taipy/gui_core/viselements.json (hunk 1 above)
Comment (FredLL-Avaiga) [reply taipy/gui_core/viselements.json]: I guess so

File: taipy/gui_core/_adapters.py [hunk 2]
```diff
@@ -348,7 +348,7 @@ def sort_key(entity: t.Union[Scenario, Cycle, Sequence, DataNode]):
     return sort_key
 
 
-@dataclass(frozen=True)
+@dataclass()
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]:
```suggestion
@dataclass
```

File: taipy/gui_core/_adapters.py [hunk 3]
```diff
@@ -368,15 +368,15 @@ def get_type(self):
         return "any"
 
 
-@dataclass(frozen=True)
+@dataclass()
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]:
```suggestion
@dataclass
```

File: taipy/gui_core/_adapters.py [hunk 4]
```diff
@@ -368,15 +368,15 @@ def get_type(self):
...
 class ScenarioFilter(_Filter):
     property_id: str
 
     def get_property(self):
         return self.property_id
 
 
-@dataclass(frozen=True)
+@dataclass()
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]:
```suggestion
@dataclass
```

File: taipy/gui_core/_adapters.py [hunk 5]
```diff
@@ -388,19 +388,25 @@ def get_property(self):
 _CUSTOM_PREFIX = "fn:"
 
 
-@dataclass(frozen=True)
+@dataclass()
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]:
```suggestion
@dataclass
```

**Review by dinhlongviolin1 — DISMISSED** [review]
Lgtm

**Review by namnguyen20999 — DISMISSED** [review]
LGTM

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Loveit!

However, the DatanodeScenarioFilter class must be renamed to DataNodeScenarioFilter for homogeneity with Core.

File: taipy/gui_core/_adapters.py (hunk 5 above)
Comment (FredLL-Avaiga) [reply taipy/gui_core/_adapters.py]: OK

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTMMMM

**Review by dinhlongviolin1 — DISMISSED** [review]
Third time the charm!

**Review by namnguyen20999 — DISMISSED** [review]
awesome !!!

**Review by namnguyen20999 — DISMISSED** [review]

**Review by FabienLelaquais — DISMISSED** [review]
🤯

**Review by FabienLelaquais — APPROVED** [review]
🎵

**Pull request merged**